docs/standard/native-interop/pinvoke-source-generation.md

```diff
@@ -0,0 +1,72 @@
+---
+title: P/Invoke source generation
+description: Learn about compile-time source generation for platform invokes in .NET.
+ms.date: 07/25/2022
+---
+
+# Source generation for platform invokes
+
+.NET 7 introduces a [source generator](../../csharp/roslyn-sdk/source-generators-overview.md) for P/Invokes which recognizes the <xref:System.Runtime.InteropServices.LibraryImportAttribute>.
+
+When a P/Invoke is defined and called like below, the built-in interop system in the .NET runtime generates an IL stub&mdash;a stream of IL instructions that is JIT-ed&mdash; at run time to facilitate the transition from managed to unmanaged.
+
+```csharp
+[DllImport("nativelib", EntryPoint = "to_lower", CharSet = CharSet.Unicode)]
+internal static extern string ToLower(string str);
+
+string lower = ToLower("StringToConvert");
+```
+
+The IL stub handles [marshalling](type-marshalling.md) of parameters and return values and calling the unmanaged code while respecting settings on <xref:System.Runtime.InteropServices.DllImportAttribute> that affect how the unmanaged code should be invoked (for example, <xref:System.Runtime.InteropServices.DllImportAttribute.SetLastError>). Since this IL stub is generated at run time, it is not available for ahead-of-time (AOT) compiler or IL trimming scenarios. Debugging the marshalling logic is also a non-trivial exercise.
```

Comment:
> Debugging the marshalling logic is also a non-trivial exercise.

Why, are there any suggestions as to how to do this?

Reply:
Because the IL stub is generated for the invocation at run time, the user can't easily step through it. For diagnostic purposes, the runtime can emit [interop events](https://docs.microsoft.com/dotnet/fundamentals/diagnostics/runtime-interop-events#ilstubgenerated-event) that contain the generated IL. I think the only instructions we have is in an [interop doc](https://github.com/AaronRobinsonMSFT/interop-doc/blob/main/doc/interop-doc.md#diagnostics) that @AaronRobinsonMSFT started (the contents of which we intend to bring to official MS docs, but haven't gotten to).